cloud/src/common/document_message.cpp

```diff
@@ -0,0 +1,369 @@
+        if (!is_splitable_field(field)) {
+            LOG(ERROR) << "Field with id " << field_id
+                       << " is not a valid message field in message "
+                       << msg.GetDescriptor()->full_name();
+            return false;
+        }
+        const google::protobuf::Reflection* reflection = msg.GetReflection();
+        if (field->is_repeated()) {
+            int field_size = reflection->FieldSize(msg, field);
+            if (field_size == 0) {
+                LOG(ERROR) << "Repeated field with id " << field_id << " is empty in message "
+                           << msg.GetDescriptor()->full_name()
+                           << ", but it is expected to be present according to the split schema.";
+                return false;
+            }
+        } else {
+            DCHECK(field->cpp_type() != google::protobuf::FieldDescriptor::CPPTYPE_MESSAGE)
+                    << "Field with id " << field_id << " is not a message field in message "
+                    << msg.GetDescriptor()->full_name();
+            if (!reflection->HasField(msg, field)) {
+                LOG(ERROR) << "Field with id " << field_id << " is not set in message "
+                           << msg.GetDescriptor()->full_name()
+                           << ", but it is expected to be present according to the split schema.";
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+// Split the fields of a message into multiple keys in the transaction.
+// Only supports repeated fields or message fields. The split fields and
+// the number of split keys are recorded in the `split_schema` field of the message.
+//
+// If a field is not set (or empty for repeated fields), it will be skipped.
+static bool document_put_split_fields(Transaction* txn, std::string_view key,
+                                      google::protobuf::Message& msg,
+                                      const std::vector<int32_t>& split_field_ids,
+                                      SplitSchemaPB* split_schema) {
+    using google::protobuf::FieldDescriptor;
+    using google::protobuf::Reflection;
+    using google::protobuf::Message;
+
+    DCHECK(ensure_fields_are_splitable(msg, split_field_ids))
+            << "The split fields must be a repeated or message field";
+
+    size_t num_put_keys = txn->num_put_keys();
+    for (int32_t field_id : split_field_ids) {
+        const FieldDescriptor* field = msg.GetDescriptor()->FindFieldByNumber(field_id);
+        const Reflection* reflection = msg.GetReflection();
+        if (field->is_repeated()) {
+            int field_size = reflection->FieldSize(msg, field);
+            if (field_size == 0) {
+                continue; // Skip empty repeated fields
+            }
+            // Split the repeated message field
+            for (int i = 0; i < field_size; ++i) {
+                std::string split_key(key);
+                encode_bytes(field->name(), &split_key);
```

Comment:
why do we accept split_field_ids but encode field_name?
there is a drawback that we cannot rename the fields in the future, or we should add a mechanism to refuse renaming splitted fields (e.g. UT)